src/reanimated2/threads.ts

```diff
@@ -153,14 +153,27 @@ if (__DEV__ && !IS_WEB) {
   }
 }
 
+function runWorkletOnJS<A extends any[], R>(
+  worklet: ComplexWorkletFunction<A, R>,
+  ...args: A
+): void {
+  // remote function that calls a worklet synchronously on the JS runtime
+  worklet(...args);
+}
+
 export function runOnJS<A extends any[], R>(
   fun: ComplexWorkletFunction<A, R>
 ): (...args: A) => void {
   'worklet';
+  if (fun.__workletHash) {
+    // if `fun` is a worklet, we schedule a call of a remote function `runWorkletOnJS`
+    // and pass the worklet as a first argument followed by original arguments
+    return (...args) => runOnJS(runWorkletOnJS<A, R>)(fun, ...args);
```

Comment:
@tjzel Thanks for the hint with `<A, R>` here! :swiniok:

Reply:
```
                                                                                
                #&%#((((/            ...            ,/(((                       
                 ##%%%(///((,(((((((((((((((((((#(((///#.                       
                 ((#####(/##((((((((///////(((((((((/(#(.                       
                 //(((%##%######((((//////////(((((((#(/.                       
                  .%##%%%%%####((((##(///////(((#((###(#                        
                     %%%%%%##%%&@@%(##(((((/(((##(&&%%                          
                     .&%%%#((###((((###((((((((///////*        .                
                      #%###(((((((((########/////////((((     .                 
                  ,#,  ###(((((((((######%#/((##(//(((##((    .                 
                 ###*. .##((((((########%%/((@&%(/((((&&#(   .  *               
               /##%%%*,. ,### #########%%(((&&&%///(((&&%(/*. .*                
              #&%%%%%%(*,..,,.*######%%%%#//&&&//////((&&//..*.                 
                %%%%&%#((/*,/,.*%%%%%%%%&&%(/////////(((###,,                   
                %%%%%&%#(/((((*,...*#%%%&&&&&&#(((/((((***/*/.#,. .#            
               ,%%%%%&&&%%########((((((((&%####%%%%%#(((((((((////*%           
               (%%%%%%&&&&&&&&&%%%%%%%####%&%%###%%%%%%#((((((((((((.           
                  &%%%%&&&&&&&&&&&&&&&&%%%%%%%%%%%%&(  (%#((((((##*             
                   &%%&&&&&&&&&&&&&&&&&&&&&&&&&&&%&      %####%,                
                    &&&&&&&&&&&&&&&&&&&&&&&&&&&&%(        /#                    
                        %&&&&&&&&&&&&&&&&&&&&#                                  
                          &%&&&&&&&&&&&&&&%&                                    
                                &%%%%&                                          
                                                                                
```